lib/copy/copy.js

```diff
@@ -2,51 +2,211 @@
+      if (err) return cb(err)
+      if (options.preserveTimestamps) {
+        return utimes(dest, srcStat.atime, srcStat.mtime, cb)
+      }
+      return cb()
+    })
+  })
+}
+
+function onDir (src, srcStat, dest, options, cb) {
+  checkDest(dest, (err, res) => {
+    if (err) return cb(err)
+    if (res === DEST_NOENT) {
+      // if dest is a subdir of src, prevent copying into itself
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      fs.mkdir(dest, srcStat.mode, err => {
+        if (err) return cb(err)
+        fs.chmod(dest, srcStat.mode, err => {
+          if (err) return cb(err)
+          return cpDir(src, dest, options, cb)
+        })
+      })
+    } else if (res === DEST_EXISTS) {
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      return cpDir(src, dest, options, cb)
+    } else if (res && typeof res !== 'number') { // dest exists and is a link
+      if (isSrcSubdir(src, res)) return cb(new Error(`Cannot copy directory '${src}' into itself '${res}'`))
+      if (src === res) return cb()
+      return cpDir(src, dest, options, cb)
     }
+  })
+}
 
-    fs.exists(dir, dirExists => {
-      if (dirExists) return ncp(src, dest, options, callback)
-      mkdir.mkdirs(dir, err => {
-        if (err) return callback(err)
-        ncp(src, dest, options, callback)
+function cpDir (src, dest, options, cb) {
+  fs.readdir(src, (err, items) => {
+    if (err) return cb(err)
+    Promise.all(items.map(item => {
+      return new Promise((resolve, reject) => {
+        copy(path.join(src, item), path.join(dest, item), options, err => {
+          if (err) reject(err)
+          else resolve()
+        })
       })
+    })).then(() => cb())
+    .catch(cb)
+  })
+}
+
+function onLink (src, dest, options, cb) {
+  fs.readlink(src, (err, resolvedSrcPath) => {
+    if (err) return cb(err)
+
+    if (options.dereference) {
+      resolvedSrcPath = path.resolve(process.cwd(), resolvedSrcPath)
+    }
+
+    checkDest(dest, (err, resolvedDestPath) => {
+      if (err) return cb(err)
+
+      if (resolvedDestPath === DEST_NOENT) {
+        // if dest is a subdir of resolved src path, prevent copying into itself
+        if (isSrcSubdir(resolvedSrcPath, dest)) {
+          return cb(new Error(`Cannot copy directory '${resolvedSrcPath}' into itself '${dest}'`))
+        }
+        return fs.symlink(resolvedSrcPath, dest, cb)
+      } else if (resolvedDestPath === DEST_EXISTS) { // dest exists but is not a link
+        if (isSrcSubdir(resolvedSrcPath, dest)) {
+          return cb(new Error(`Cannot copy directory '${resolvedSrcPath}' into itself '${dest}'`))
+        }
+        // if src points to dest
+        if (resolvedSrcPath === dest) return cb()
+        return copy(resolvedSrcPath, dest, options, cb)
+      } else if (resolvedDestPath && typeof resolvedDestPath !== 'number') { // dest is a link
```

Comment:
Again here also, when dest exists (regular and symlink), should go with the same pattern of checking `options.overwrite` and operates accordingly.